### PR TITLE
Support of Digest extern in P4TC

### PR DESCRIPTION
**Pull request by komaljai**

### CODE REVIEW

File: backends/tc/backend.cpp [hunk 1]
```diff
@@ -841,7 +841,8 @@ safe_vector<const IR::TCKey *> ConvertToBackendIR::processExternConstructor(
             } else {
                 /* If a parameter is not annoated by tc_init or tc_numel then it is emitted as
                 constructor parameters.*/
-                IR::TCKey *key = new IR::TCKey(0, parameter->type->width_bits(),
+                cstring ptype = "bit"_cs + std::to_string(parameter->type->width_bits());
```
Comment (asl) [inline backends/tc/backend.cpp:844]: consider using abseil's StrCat

File: backends/tc/backend.cpp [hunk 2]
```diff
@@ -910,8 +911,9 @@ safe_vector<const IR::TCKey *> ConvertToBackendIR::processCounterControlPathKeys
             auto temp_keys = HandleTypeNameStructField(field, extn, decl, kId, annoName);
             keys.insert(keys.end(), temp_keys.begin(), temp_keys.end());
         } else {
-            IR::TCKey *key =
-                new IR::TCKey(kId++, field->type->width_bits(), field->toString(), annoName, true);
+            cstring ptype = "bit"_cs + std::to_string(field->type->width_bits());
```
Comment (asl) [inline backends/tc/backend.cpp:914]: ditto

File: backends/tc/backend.cpp [hunk 3]
```diff
@@ -942,8 +944,9 @@ safe_vector<const IR::TCKey *> ConvertToBackendIR::processExternControlPath(
                 auto temp_keys = HandleTypeNameStructField(field, extn, decl, kId, annoName);
                 keys.insert(keys.end(), temp_keys.begin(), temp_keys.end());
             } else {
-                IR::TCKey *key = new IR::TCKey(kId++, field->type->width_bits(), field->toString(),
-                                               annoName, true);
+                cstring ptype = "bit"_cs + std::to_string(field->type->width_bits());
```
Comment (asl) [inline backends/tc/backend.cpp:947]: ditto

File: backends/tc/backend.cpp [hunk 4]
```diff
@@ -965,13 +968,27 @@ safe_vector<const IR::TCKey *> ConvertToBackendIR::HandleTypeNameStructField(
             /* If 'T' is of Type_Struct, extract all fields of structure*/
             if (auto param_struct = param_val->to<IR::Type_Struct>()) {
                 for (auto f : param_struct->fields) {
-                    IR::TCKey *key =
-                        new IR::TCKey(kId++, f->type->width_bits(), f->toString(), annoName, true);
+                    cstring ptype = "bit"_cs + std::to_string(f->type->width_bits());
```
Comment (asl) [inline backends/tc/backend.cpp:971]: ditto

File: backends/tc/backend.cpp [hunk 5]
```diff
@@ -965,13 +968,27 @@ safe_vector<const IR::TCKey *> ConvertToBackendIR::HandleTypeNameStructField(
...
+                        }
+                    }
+                    IR::TCKey *key = new IR::TCKey(kId++, f->type->width_bits(), ptype,
+                                                   f->toString(), annoName, true);
                     keys.push_back(key);
                 }
             } else {
-                IR::TCKey *key = new IR::TCKey(kId++, param_val->width_bits(), field->toString(),
-                                               annoName, true);
+                cstring ptype = "bit"_cs + std::to_string(param_val->width_bits());
```
Comment (asl) [inline backends/tc/backend.cpp:983]: ditto

File: backends/tc/backend.cpp (hunk 1 above)
Comment (komaljai) [reply backends/tc/backend.cpp:844]: addressed

File: backends/tc/backend.cpp [hunk 6]
```diff
@@ -910,8 +911,9 @@ safe_vector<const IR::TCKey *> ConvertToBackendIR::processCounterControlPathKeys
             auto temp_keys = HandleTypeNameStructField(field, extn, decl, kId, annoName);
             keys.insert(keys.end(), temp_keys.begin(), temp_keys.end());
         } else {
-            IR::TCKey *key =
-                new IR::TCKey(kId++, field->type->width_bits(), field->toString(), annoName, true);
+            cstring ptype = absl::StrCat("bit", std::to_string(field->type->width_bits()));
```
Comment (asl) [inline backends/tc/backend.cpp:914]: You do not need `to_string`. Consider reading https://abseil.io/docs/cpp/guides/strings#abslstrcat-and-abslstrappend-for-string-concatenation

File: backends/tc/tc.def [hunk 7]
```diff
@@ -446,7 +448,7 @@ class TCKey {
         emitValue = true;
     }
     toString {
-        std::string tckeyInstance = " " + keyAttribute + " " + keyName + " ptype bit" + Util::toString(bitwidth);
+        std::string tckeyInstance = " " + keyAttribute + " " + keyName + " ptype " + type;
```
Comment (asl) [inline backends/tc/tc.def:451]: while you are here, I would suggest you change to `StrCat` here and to `StrAppend` two instances below. Not a big deal, but it is good to make cleanup in parallel :)

**Comment by asl** [comment]
String changes are ok to me (see small suggestions).

**Review by Sosutha — APPROVED** [review]
Looks fine

**Review by fruffy — CHANGES_REQUESTED** [review]
Please address @asl's last request before merging

File: backends/tc/tc.def (hunk 7 above)
Comment (komaljai) [reply backends/tc/tc.def:446]: addressed

File: backends/tc/tc.def [hunk 8]
```diff
@@ -79,61 +80,60 @@ class TCActionParam {
         std::string paramType = "";
         switch(dataType) {
             case TC::BIT_TYPE :
-                paramType += "__u" + Util::toString(bitSize);
+                absl::StrAppend(&paramType, "__u", bitSize);
```
Comment (asl) [inline backends/tc/tc.def:83]: No need for appends here, `paramType` is empty above, do a direct assignment.

File: backends/tc/tc.def [hunk 9]
```diff
@@ -79,61 +80,60 @@ class TCActionParam {
...
                 break;
         } 
         return paramType;
     }
     toString {
         std::string tcActionParam = "";
-        tcActionParam += "\n\tparam ";
-        tcActionParam += paramName;
-        tcActionParam += " type ";
+        absl::StrAppend(&tcActionParam, "\n\tparam ", paramName.c_str());
```
Comment (asl) [inline backends/tc/tc.def:111]: Use `string_view()` method each time you'd need to pass `cstring`'s underlying buffer down to `string_view` argument

File: backends/tc/tc.def [hunk 10]
```diff
@@ -79,61 +80,60 @@ class TCActionParam {
...
+        absl::StrAppend(&tcActionParam, " type ");
```
Comment (asl) [inline backends/tc/tc.def:112]: why multiple calls? Use a single `absl::StrCat` call and direct assignment, no need to append to empty string

File: backends/tc/tc.def [hunk 11]
```diff
@@ -156,9 +156,9 @@ class TCDefaultActionParam {
     }
     toString {
         std::string tcActionParam = "";
-        tcActionParam += " " + paramDetail->paramName;
+        absl::StrAppend(&tcActionParam, " ", paramDetail->paramName.c_str());
```
Comment (asl) [inline backends/tc/tc.def:159]: see above

File: backends/tc/tc.def [hunk 12]
```diff
@@ -174,7 +174,7 @@ class TCAction {
             return actionName;
         }
         std::string tcAction = pipelineName.string();
```
Comment (asl) [inline backends/tc/tc.def:176]: Use `StrCat`, use `string_view` instead of `string`

File: backends/tc/tc.def [hunk 13]
```diff
@@ -196,21 +196,18 @@ class TCAction {
     }
     toString {
         std::string tcAction = "";
-        tcAction += "\n$TC p4template create action/";
-        tcAction += pipelineName;
-        tcAction += "/" + actionName;
+        absl::StrAppend(&tcAction, "\n$TC p4template create action/");
```
Comment (asl) [inline backends/tc/tc.def:199]: ditto

File: backends/tc/tc.def [hunk 14]
```diff
@@ -196,21 +196,18 @@ class TCAction {
...
             for (auto actParam : actionParams) {
-                tcAction += " \\";
-                tcAction += actParam->toString();
+                absl::StrAppend(&tcAction, " \\", actParam->toString().c_str());
             }
         }
-        tcAction += "\n$TC p4template update action/";
-        tcAction += pipelineName;
-        tcAction += "/" + actionName + " state active";
+        absl::StrAppend(&tcAction, "\n$TC p4template update action/");
```
Comment (asl) [inline backends/tc/tc.def:209]: ditto

File: backends/tc/tc.def [hunk 15]
```diff
@@ -226,7 +223,7 @@ class TCEntry {
     toString {
         std::string tcEntry = "";
         for(auto k : keys) {
-            tcEntry += " " + k.first + " " + k.second;
+            absl::StrAppend(&tcEntry, " ", k.first.c_str(), " ", k.second.c_str());
```
Comment (asl) [inline backends/tc/tc.def:226]: Use `StrJoin` with (likely) custom formatter instead of loop

File: backends/tc/tc.def [hunk 16]
```diff
@@ -349,75 +346,73 @@ class TCTable {
...
         std::string tcTable = "";
-        tcTable += "\n$TC p4template create table/";
-        tcTable += pipelineName + "/" + controlName + "/" + tableName + " \\";
-        tcTable += "\n\ttblid " + Util::toString(tableID) + " \\";
-        tcTable += "\n\ttype " + printMatchType(matchType) + " \\";
-        tcTable += "\n\tkeysz " + Util::toString(keySize);
-        tcTable += " nummasks " + Util::toString(numMask);
-        tcTable += " permissions " + Util::toString(permissions);
-        tcTable += " tentries " + Util::toString(tableEntriesCount);
+        absl::StrAppend(&tcTable, "\n$TC p4template create table/");
```
Comment (asl) [inline backends/tc/tc.def:349]: see above

File: backends/tc/tc.def [hunk 17]
```diff
@@ -349,75 +346,73 @@ class TCTable {
...
         }
         if (const_entries.size() != 0) {
             for (auto entry : const_entries) {
-                tcTable += "\n$TC p4template create table/" + pipelineName
-                            + "/" + controlName + "/" + tableName
-                            + " entry" + entry->toString();
-                tcTable += " permissions 0x1024";
-                tcTable += " action " + pipelineName
-                            + "/" + controlName + "/" + entry->getActionName();
+                absl::StrAppend(&tcTable, "\n$TC p4template create table/", pipelineName.c_str(),
```
Comment (asl) [inline backends/tc/tc.def:410]: ditto

File: backends/tc/tc.def [hunk 18]
```diff
@@ -446,12 +443,12 @@ class TCKey {
         emitValue = true;
     }
     toString {
-        std::string tckeyInstance = " " + keyAttribute + " " + keyName + " ptype bit" + Util::toString(bitwidth);
+        std::string tckeyInstance = absl::StrCat(" ", keyAttribute.c_str(), " ", keyName.c_str(), " ptype ", type.c_str());
```
Comment (asl) [inline backends/tc/tc.def:446]: what happened with `bitwidth`? See above about `string_view`

File: backends/tc/tc.def [hunk 19]
```diff
@@ -498,28 +495,25 @@ class TCExternInstance {
         numelemns = ne;
     } 
     toString {
-        std::string tcExternInstance = "";
-        tcExternInstance += instanceName + " instid " + Util::toString(instanceID) + " \\";
-        
+        std::string tcExternInstance = absl::StrCat(instanceName.c_str(), " instid ");
```
Comment (asl) [inline backends/tc/tc.def:498]: see above: use `StrCat`

File: backends/tc/tc.def [hunk 20]
```diff
@@ -555,20 +549,16 @@ class TCExtern {
         return nullptr;
     }
     toString {
-        std::string tcExtern = "";
-        tcExtern += "\n$TC p4template create extern/";
-        tcExtern += "root/" + externName;
-        tcExtern += " extid " + externID;
-        tcExtern += " numinstances " + Util::toString(numinstances);
-        tcExtern += " tc_acl " + acl_permisson;
+        std::string tcExtern = absl::StrCat("\n$TC p4template create extern/", "root/", externName.c_str());
```
Comment (asl) [inline backends/tc/tc.def:552]: ditto

File: backends/tc/tc.def [hunk 21]
```diff
@@ -623,37 +613,37 @@ class TCPipeline {
     }
     toString {
         std::string tcCode = "#!/bin/bash -x\n";
-        tcCode += "\nset -e\n";
-        tcCode += "\nTC=\"tc\"";
-        tcCode += "\n$TC p4template create pipeline/" + pipelineName;
-        tcCode += " numtables ";
-        tcCode += Util::toString(numTables);
+        absl::StrAppend(&tcCode, "\nset -e\n");
```
Comment (asl) [inline backends/tc/tc.def:616]: ditto

File: backends/tc/tc.def [hunk 22]
```diff
@@ -623,37 +613,37 @@ class TCPipeline {
...
             for (auto t : tableDefs) {
-                tcCode += "\n" + t->toString();
+                absl::StrAppend(&tcCode, "\n", t->toString().c_str());
             }
         }
         if (preaction != nullptr) {
-            tcCode += "\n" + preaction->toString();
-            tcCode += "\n$TC p4template update pipeline/" + pipelineName
-                      + " preactions action " + pipelineName + "/preaction";
+            absl::StrAppend(&tcCode, "\n", preaction->toString().c_str());
```
Comment (asl) [inline backends/tc/tc.def:637]: ditto

File: backends/tc/tc.def [hunk 23]
```diff
@@ -623,37 +613,37 @@ class TCPipeline {
...
+            absl::StrAppend(&tcCode, "\n$TC p4template update pipeline/", pipelineName.c_str(),
+                      " preactions action ", pipelineName.c_str(), "/preaction");
         }
         if (postaction != nullptr) {
-            tcCode += "\n" + postaction->toString();
-            tcCode += "\n$TC p4template update pipeline/" + pipelineName
-                      + " postactions action " + pipelineName + "/postaction";
+            absl::StrAppend(&tcCode, "\n", postaction->toString().c_str());
```
Comment (asl) [inline backends/tc/tc.def:642]: ditto

**Review by asl — CHANGES_REQUESTED** [review]
Looks like the code changes were purely mechanical. The code will benefit a lot, if it will be changed to do string operations in better way

File: backends/tc/tc.def (hunk 18 above)
Comment (komaljai) [reply backends/tc/tc.def:446]: Changing bitwidth to ptype was original PR change.

File: backends/tc/tc.def (hunk 10 above)
Comment (komaljai) [reply backends/tc/tc.def:112]: I kept multiple lines for better readability merged it to one however now.

Comment (asl) [reply backends/tc/tc.def:112]: Then looks like there is a misunderstanding of why we are using these string routines and what are their purpose. Note that every `append` call essentially does a memory reallocation that incurs some cost. These costs might be elevated if these is used in e.g. hot loop

File: backends/tc/tc.def (hunk 8 above)
Comment (asl) [reply backends/tc/tc.def:83]: Please do not resolve unaddressed comments

File: backends/tc/tc.def (hunk 11 above)
Comment (asl) [reply backends/tc/tc.def:159]: Unaddressed

**Review by asl — CHANGES_REQUESTED** [review]
Looks like you decide to ignore all my comments just resolving all of them. I doubt this is how review process work.

I'm leaving this for @fruffy

File: backends/tc/tc.def (hunk 18 above)
Comment (asl) [reply backends/tc/tc.def:446]: Ok. Then why have you resolved it without addressing `string_view` comment?

**Comment by komaljai** [comment]
Hi @asl , I did address all comments, as they were many minor comments, I tried to mark it resolve as I fixed. (my bad). The unit tests were failing in the end, due to which I couldn't push yesterday. I have pushed them now.

**Review by fruffy — APPROVED** [review]
Seem there was a misunderstanding here but I believe most complaints are resolved now.

**Comment by komaljai** [comment]
> Seem there was a misunderstanding here but I believe most complaints are resolved now.

Thanks @fruffy

**Pull request merged**